net_connection/src/net_connection_thread.rs

```diff
@@ -0,0 +1,178 @@
+use super::NetResult;
+
+use super::{
+    net_connection::{NetConnection, NetHandler, NetWorkerFactory},
+    protocol::Protocol,
+};
+
+use std::{thread, time};
+
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    mpsc, Arc,
+};
+
+/// a NetConnection instance that is managed on another thread
+#[derive(Debug)]
+pub struct NetConnectionThread {
+    keep_running: Arc<AtomicBool>,
+    send_channel: mpsc::Sender<Protocol>,
+    thread: thread::JoinHandle<()>,
+}
+
+impl NetConnection for NetConnectionThread {
+    /// send a message to the worker within this NetConnectionThread instance
+    fn send(&mut self, data: Protocol) -> NetResult<()> {
+        self.send_channel.send(data)?;
+        Ok(())
+    }
+}
+
+impl NetConnectionThread {
+    /// stop (join) the worker thread
+    pub fn stop(self) -> NetResult<()> {
+        self.keep_running.store(false, Ordering::Relaxed);
+        match self.thread.join() {
```

Comment:
You can make this a simple `if` using:

```rust
if self.thread.join().is_err() {
    bail!("NetConnectionThread failed to join on stop() call");
}

Ok(())
```

But that's really personal preference